labgrid/driver/bareboxdriver.py

```diff
@@ -79,25 +91,21 @@ def _run(self, cmd: str, *, timeout: int = 30, codec: str = "utf-8", decodeerror
         Returns:
             Tuple[List[str],List[str], int]: if successful, None otherwise
         """
-        # FIXME: use codec, decodeerrors
-        marker = gen_marker()
-        # hide marker from expect
-        hidden_marker = f'"{marker[:4]}""{marker[4:]}"'
-        cmp_command = f'''echo -o /cmd {shlex.quote(cmd)}; echo {hidden_marker}; sh /cmd; echo {hidden_marker} $?;'''  # pylint: disable=line-too-long
         if self._status == 1:
-            self.console.sendline(cmp_command)
-            _, _, match, _ = self.console.expect(
-                rf'{marker}(.*){marker}\s+(\d+)\s+.*{self.prompt}',
-                timeout=timeout)
-            # Remove VT100 Codes and split by newline
-            data = self.re_vt100.sub('', match.group(1).decode('utf-8')).split('\r\n')[1:-1]
-            self.logger.debug("Received Data: %s", data)
-            # Get exit code
-            exitcode = int(match.group(2))
-            return (data, [], exitcode)
+            with self._popen(cmd) as p:
+                output = p.read(timeout=timeout)
+                # Remove VT100 Codes and split by newline
+                data = self.re_vt100.sub('', output.decode('utf-8')).split('\r\n')[1:-1]
+                self.logger.debug(f"Received Data: {data}")
```

Comment:
log functions should not use f-strings, but rather lazy % formatting. There are multiple occurrences of this spread in this PR.

Reply:
Yes, there is a reason. Here is a short wrap-up: https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

TL;DR: formatting is only done if required (i.e. the log level is reached) and errors are handled gracefully.

Reply:
Reading the article has confused me more; IIUC, the article is suggesting that logging should use lazy evaluation by passing format arguments as parameters (_not_ using `%`), in which case the log message would be something like `self.logger.debug("Received data: %s", data)`. However, it seems you are suggesting I use `%` instead, which is _not_ a lazy evalutation and therefore not superior (at least in terms of what the article is discussing) from f-strings.

I _think_ what you want to do is either use actual lazy evaluation for log messages, or non-lazy f-strings for consistency with everywhere else f-strings are used

Reply:
With "% formatting" I mean `logger.debug("%s", data)`.

Reply:
OK, I will do that